### PR TITLE
TLS 1.3: Upstream misc fix from prototype

### LINKED ISSUE #6311: Merge of development TLS 1.3 Q3 work into the TLS 1.3 prototype
The TLS 1.3 work done in CYQ3 (PSK support on client and server side) needs to be merged (in the git sense) into the [TLS 1.3 prototype](https://github.com/hannestschofenig/mbedtls): up to the merge of PR #6482 (commit 04e2133f4).

After the merge, the following tests must still run successfully:
1. all "test_tls13*" all.sh components
2. the test_full_cmake_gcc_asan all.sh component (build with both TLS 1.2 and TLS 1.3)
3. unit tests and compat.sh in default configuration (ssl-opt.sh not running successfully in that configuration currently)

The merge of TLS 1.3 upstream back to the prototype has been done in subsequent merges of TLS 1.3 related merge commits. List of the TLS 1.3 related merge commits since the last merge of tag 3.2.1 into the prototype (to be checked) to be merged as part of addressing this issue:

04e2133f4 Merge pull request #6482 from ronald-cron-arm/tls13-misc                         
77e15e8a2 Merge pull request #6460 from xkqian/tls13_add_early_data_preparatory            
2012361ae Merge pull request #6448 from ronald-cron-arm/tls13-kex-build-options            
49e418481 Merge pull request #6299 from xkqian/tls13_add_servername_check                  
78317c832 Merge pull request #6327 from yuhaoth/pr/tls13-psk-after-session-tickets         
77c691f09 Merge pull request #6194 from xkqian/tls13_add_psk_client_cases                  
2c282c9bd Merge pull request #6180 from yuhaoth/pr/add-tls13-multiple-session-tickets      
07ba2be20 Merge pull request #6304 from yuhaoth/pr/exclude-pre_shared_key-from-hrr-msg     
be0224aef Merge pull request #6167 from yuhaoth/pr/finalize-tls13-session-tickets          
62e24ba18 Merge pull request #6260 from yuhaoth/pr/add-multiple-pre-config-psks            
208257b39 Merge pull request #6259 from yuhaoth/pr/add-psk_ephemeral-possible-group-tests  
e00d6d6b5 Merge pull request #6135 from yuhaoth/pr/tls13-finalize-external-psk-negotiation 
f3f6b0a5c Merge pull request #6123 from yuhaoth/pr/finialize-tls13-serialize_session_save_load
322a7a19e Merge pull request #6155 from yuhaoth/pr/add-any-all-configs-enabled             
953ce3962 Merge pull request #5971 from yuhaoth/pr/add-rsa-pss-rsae-for-tls12              
e579ece30 Merge pull request #6087 from yuhaoth/pr/add-tls13-serialize_session_save_load   
340c559cb Merge pull request #6079 from yuhaoth/pr/add-tls13-parse-pre_shared_key_offered_psks
4beb870fa Merge pull request #6064 from xkqian/tls13_add_psk                    
32578b3bd Merge pull request #6069 from yuhaoth/pr/add-tls13-write-new-session-ticket      
892747015 Merge pull request #6112 from ronald-cron-arm/tls13-negotiated-key-exchange-mode 
90ae438cf Merge pull request #5913 from yuhaoth/pr/add-tls13-new-session-ticket 
d5b1eb51d Merge pull request #6078 from yuhaoth/pr/add-tls13-paser-psk-kex-mode-ext

Thus the first merge should go something like:
"git merge d5b1eb51d"

After the merge of an upstream merge commit into the prototype, check that all the above mentioned tests still pass.

As examples of what has been done so far, the two following branches contain the merges up to v3.1.0 and v3.2.1:
https://github.com/ronald-cron-arm/mbedtls/tree/merge_5114_to_v3_1_0 
https://github.com/ronald-cron-arm/mbedtls/tree/merge_5539_to_v3_2_1

Note the presence of commits (like 07dd3e9011e15b5e04f8b6d3c3f3172d42065e35 in merge_5539_to_v3_2_1) between two merges of upstream merge commits to the prototype to simplify the diff between the merge results. This is particularly relevant when an upstream merge commit adds new functions that exist already in the prototype: the prototype refactoring commit then try to align in the prototype the position in the files of the aforementioned functions as well as their naming. 

Justification
We want to keep the TLS 1.3 support in Mbed TLS and the TLS 1.3 prototype to converge.

### PR DESCRIPTION
**Pull request by yuhaoth**
## Description

This PR include some small fix and some better code from `prototype`

fix #6311 

## Gatekeeper checklist

- [ ] **changelog** provided, or not required
- [ ] **backport** done, or not required
- [ ] **tests** provided, or not required

## Notes for the submitter

Please refer to the [contributing guidelines](../CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

File: library/ssl_tls.c [hunk 1]
```diff
@@ -3477,6 +3479,8 @@ static unsigned char ssl_serialized_session_header[] = {
  *
  *      case MBEDTLS_SSL_VERSION_TLS1_2:
  *        serialized_session_tls12 data;
+ *      case MBEDTLS_SSL_MINOR_VERSION_4:
```
Comment (lpy4105) [inline library/ssl_tls.c:3482]: Should this be `MBEDTLS_SSL_VERSION_TLS1_3` ?
Should we add the following line under line 3476?
```C
 *                         // - TLS 1.2 (3)
```

**Review by lpy4105 — COMMENTED** [review]
Several minor comments.

File: library/ssl_tls.c [hunk 2]
```diff
@@ -3189,12 +3189,14 @@ size_t mbedtls_ssl_get_input_max_frag_len( const mbedtls_ssl_context *ssl )
     size_t max_len = MBEDTLS_SSL_IN_CONTENT_LEN;
     size_t read_mfl;
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2)
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply library/ssl_tls.c:3192]: TLS 1.3 does not include `MBEDTLS_SSL_SERVER_HELLO_DONE` .  And the state machine in TLS 1.3 does not use `ssl->state++` which is bad and unreadable.

**Review by lpy4105 — COMMENTED** [review]
LGTM, will give approval after CI pass.

**Comment by yuhaoth** [comment]
Rebase to trigger CI test again

**Review by lpy4105 — APPROVED** [review]
Internal CI pass.

File: library/ssl_tls.c [hunk 3]
```diff
@@ -4748,14 +4762,18 @@ void mbedtls_ssl_free( mbedtls_ssl_context *ssl )
...
-        mbedtls_ssl_session_free( ssl->session_negotiate );
-
         mbedtls_free( ssl->handshake );
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_2)
+        mbedtls_ssl_transform_free( ssl->transform_negotiate );
         mbedtls_free( ssl->transform_negotiate );
+#endif
+
+        mbedtls_ssl_session_free( ssl->session_negotiate );
```
Comment (mprse) [inline library/ssl_tls.c:4780]: This looks like preexisting bug fix. Memory for session `hostname`, `ticket`, `cert` was not freed before entire session was freed. It is strange that memory leaks were not detected on CI.

**Review by mprse — APPROVED** [review]
LGTM.

Left one comment for clarification.

**Pull request merged**